feat(devtools): adding router graph implementation

sumitarora opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application/infrastructure changes
- [ ] Other... Please describe:

## What is the new behaviour?
* New router graph implementation for Angular Devtools to inspect application routes
* Add new functionality to export debug APIs using `IPublishUtils` global 
* Export new `getLoadedRoutes` function to get child routes while parsing router config.

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

<img width="1892" alt="image" src="https://github.com/angular/angular/assets/198247/fc43691f-13a3-404d-9895-807f393405a8">

<img width="1843" alt="image" src="https://github.com/angular/angular/assets/198247/28a4ae46-4a5b-468c-90c9-f769a7bbf73f">

sumitarora committed:
refactor(devtools): adding router demo application

- added demo application containing different route variations

sumitarora committed:
refactor(devtools): render router graph for devtools

- added base router graph rendering router data

sumitarora committed:
refactor(devtools): fetch router instance from root injector

- fetch the router data using the root router instance

eneajaho commented:
Hello @sumitarora 
Great job. 
Any idea if we can add two modes on the router? 

One like a tree, and one like a list. 

Some idea from vue devtools

And we can also add some cool features like: 
- all the guards that are being used,
- all the route data
- how much time it takes the route to be opened and things like that. 

![image](https://github.com/angular/angular/assets/25394362/4ebff027-6c45-4835-899d-e38252f3d63a)
![image](https://github.com/angular/angular/assets/25394362/71d06344-76cf-4df0-8746-769bf784ccfb)

sumitarora commented:
Hey @eneajaho we can add it in future PRs once we have merged this.

AleksanderBodurri commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -211,10 +213,15 @@ const getNestedPropertiesCallback =
+    initializeOrGetDirectiveForestHooks().getIndexedDirectiveForest(),
+    ngDebugDependencyInjectionApiIsSupported(),
+  );
+  const rootInjector = forest[0].resolutionPath?.find((i) => i.name === 'Root');
```
If `ngDebugDependencyInjectionApiIsSupported()` is false then there will be no resolution paths in the forest nodes. We should handle this case by doing `forest[0].resolutionPath?.find?.(...)` or the suggestion below

```suggestion
  const rootInjector = (forest[0].resolutionPath ?? []).find((i) => i.name === 'Root');
```

AleksanderBodurri commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -405,29 +426,30 @@ const logProvider = (
+    providerValue = injector.get(provider[0].token, null, {optional: true});
+  }
 
+  if (returnValue) {
```
Lets split out the logic that grabs `providerValue` into a helper function so that we don't need to add this boolean arg here

AleksanderBodurri commented on devtools/projects/ng-devtools-backend/src/lib/router-tree.ts:
```diff
@@ -14,55 +14,88 @@ type Routes = any;
+    const childDescendents: [any] = loadedRoutes || child.children;
 
     const pathFragment = child.outlet ? `(${child.outlet}:${child.path})` : child.path;
+    const routPath = `${parentPath ? parentPath : ''}/${pathFragment}`.split('//').join('/');
```
typo: `routePath`

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/devtools-tabs.component.ts:
```diff
@@ -65,9 +65,11 @@ export class DevToolsTabsComponent implements OnInit, AfterViewInit {
 
   @Output() frameSelected = new EventEmitter<Frame>();
   @ViewChild(DirectiveExplorerComponent) directiveExplorer!: DirectiveExplorerComponent;
+  @ViewChild(RouterTreeComponent) routerTreeComponent!: RouterTreeComponent;
```
Do we need this ViewChild here?

AleksanderBodurri commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.scss:
```diff
@@ -28,38 +49,104 @@
 }
 
+.selected-route {
+  background-color: #ebf1fb !important;
```
Let's target these with specificity instead of `!important`

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree-visualizer.ts:
```diff
@@ -0,0 +1,333 @@
+
+  override snapToNode(node: RouterTreeD3Node, scale = 1): void {
+    const svg = this.d3.select(this._containerElement);
+    // const halfWidth = this._containerElement.clientWidth / 2;
```
Nit: Can we drop this?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree-visualizer.ts:
```diff
@@ -0,0 +1,333 @@
+
+  constructor(
+    private _containerElement: HTMLElement,
+    private _graphElement: HTMLElement,
```
Style: I suggest dropping the leading `_` for `private` symbols. TypeScript already enforces visibility and we don't need to worry about JS interop here.

I think we've been using `_` in other places of DevTools, and it's probably not important enough to worry about going back and changing, but maybe just for we just shouldn't add it to new code?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree-visualizer.ts:
```diff
@@ -0,0 +1,333 @@
+
+    const size = 20;
+
+    svg
```
Suggestion: Some comments about what each of these are drawing would be really helpful. I _think_ this is drawing the legend in the top-left, but I'm not entirely sure.

If so, separating this out into a `renderLegend` function might be useful for compartmentalization as well.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree-visualizer.ts:
```diff
@@ -0,0 +1,333 @@
+        return `node`;
+      })
+      .on('click', (pointerEvent: PointerEvent, node: RouterTreeD3Node) => {
+        this.nodeClickListeners.forEach((listener) => listener(pointerEvent, node));
```
Style: Prefer `for-of` loops - https://techhub.social/@develwithoutacause/111819784287279149

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree-visualizer.ts:
```diff
@@ -0,0 +1,333 @@
+        const label =
+          (showFullPath
+            ? node.data.path
+            : node.data.path.replace(node.parent?.data.path || '', '')) || '';
```
Style: Prefer `??` over `||` when used to assign fallback / default values. `??` applies only to `null` and `undefined` which is much more intuitive than `||` which applies to any falsey value.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree-visualizer.ts:
```diff
@@ -0,0 +1,333 @@
+        }
+        return nodeClass;
+      })
+      .html((node: RouterTreeD3Node) => {
```
Question: I'm a little worried about potential XSS here. If an attacker compromises the `label` parameter, they could convert it into HTML on the page. I'm not sure if d3 does any sanitiziation here, but this feels unnecessarily risky. Could we set the text direct via some equivalent of `.textContent` and use a class/style to configure the bold/underline styling?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree-visualizer.ts:
```diff
@@ -0,0 +1,333 @@
+
+  onNodeMouseout(cb: (pointerEvent: PointerEvent, node: U) => void): void {
+    this.nodeMouseoutListeners.push(cb);
+  }
```
Question: Are these handlers actually used anywhere? I don't see them called.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.html:
```diff
@@ -1,8 +1,16 @@
 <div class="router-tree-wrapper">
-  <div class="no-routes" [hidden]="routes?.length !== 0">
-    <h1>There are no routes to display.</h1>
+  <div>
```
Nit: Do we need this `<div>` if it doesn't apply any classes or functionality?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.html:
```diff
@@ -1,8 +1,16 @@
+      placeholder="Search routes"
+      class="filter-input"
+    />
+    <mat-checkbox (change)="togglePathSettings()"> Show Full Path </mat-checkbox>
```
Nit: Drop leading/trailing whitespace like:

```html
<mat-checkbox>Show Full Path</mat-checkbox>
```

It's not very important, but HTML is typically whitespace-sensitive to this kind of thing, except that Angular disables that by default `@Component({ preserveWhitespace: /* ... */ })`. Having the whitespace here makes it look like we're intending this to be rendered, when it actually won't. So eliminating this whitespace should align the code with your actual intent as a developer.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.scss:
```diff
@@ -28,38 +49,111 @@
+}
+
 :host {
   ::ng-deep {
```
Question: Do all these need to be inside an `::ng-deep`? Feels like most of them are our own classes.

I'd definitely like to minimize usage of this as much as we reasonably can.

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.scss:
```diff
@@ -28,38 +49,111 @@
+    }
+
+    .routesTable tr:nth-child(even) {
+      // background-color: #f2f2f2;
```
Nit: Some more comments in here which look like they can be removed?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts:
```diff
@@ -6,150 +6,79 @@
+      this.filterRegex = new RegExp(this.searchString.toLowerCase() || '.^');
+    } else {
+      this.filterRegex = new RegExp('.^');
+    }
```
Suggestion: Nullish coalescing can clean this up a bit and reduce duplication of the default regex.

```typescript
this.filterRegex = new RegExp(this.searchString?.toLowerCase() ?? '.^');
```

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts:
```diff
@@ -6,150 +6,79 @@
+    if (this.searchString) {
+      this.filterRegex = new RegExp(this.searchString.toLowerCase() || '.^');
+    } else {
+      this.filterRegex = new RegExp('.^');
```
Question: What exactly does `.^` match? I don't think I've seen a `^` used in this way. It doesn't seem like it's "start of line" and it's not `[^a]`. What are we intending here?

dgp1130 commented on devtools/projects/ng-devtools/src/lib/devtools-tabs/router-tree/router-tree.component.ts:
```diff
@@ -6,150 +6,79 @@
-
-    const svgRect = this.svgContainer.nativeElement.getBoundingClientRect();
-    const gElRect = this.g.nativeElement.getBoundingClientRect();
+  searchRoutes($event: any) {
```
Nit: Would `Event` be a more appropriate type here?

dgp1130 commented on devtools/projects/protocol/src/lib/messages.ts:
```diff
@@ -203,14 +203,21 @@ export interface UpdatedStateData {
-  data: any;
+  name?: string;
+  hash?: string | null;
+  specificity?: string | null;
```
Question: What is the semantic difference between `hash: undefined`, and `hash: null`? Same for `specificity`?

Do those mean different things or should we consolidate on a single representation of "no value".

dgp1130 commented on packages/core/src/render3/util/global_utils.ts:
```diff
@@ -103,22 +95,44 @@ export type GlobalDevModeUtils = {
+ * Publishes the given function from core to `window.ng` so that it can be
  * used from the browser console when an application is not in production.
  */
+<<<<<<< HEAD
```
Bug: Looks like a merge conflict snuck in?

dgp1130 commented on packages/core/src/render3/util/global_utils.ts:
```diff
@@ -12,23 +12,8 @@ import {setProfiler} from '../profiler';
 import {isSignal} from '../reactivity/api';
```
Suggestion: Should we move `@angular/core` and/or `@angular/router` changes into their own commit(s)? That might make the FW side of the review easier here.

dgp1130 reviewed (commented):
Two commits in right now and left some comments. Mostly minor nitpicky stuff, feel free to ignore or push back on any of the stylistic things to disagree with.

dgp1130 commented on packages/core/src/core.ts:
```diff
@@ -98,6 +98,7 @@ export {
 } from './render3/ng_module_ref';
 export {createComponent, reflectComponentType, ComponentMirror} from './render3/component';
 export {isStandalone} from './render3/definition';
+export {publishExternalGlobalUtil} from './render3/util/global_utils';
```
Suggestion: I think we'll want this to be a private export.

```typescript
export {publishExternalGlobalUtil as ɵpublishExternalGlobalUtil} from '...';
```

dgp1130 commented on packages/router/src/router_devtools.ts:
```diff
@@ -0,0 +1,21 @@
+  }
+}
+
+function getLoadedRoutes(route: Route): Route[] | undefined {
```
Question: Is there a semantic difference between a return value of `[]` and `undefined`? Could we just default to `[]` if `._loadedRoutes` is `undefined` and would that reduce the complexity here?

dgp1130 reviewed (commented):
Left some more thoughts on the second two commits. Hopefully this is useful feedback and can help push things forward a bit.

sumitarora commented on devtools/projects/ng-devtools-backend/src/lib/router-tree.ts:
```diff
@@ -14,55 +14,88 @@ type Routes = any;
+    const loadedRoutes =
+      (window as any).ng &&
+      (window as any).ng.getLoadedRoutes &&
+      (window as any).ng.getLoadedRoutes(child as any);
```
In devtools backend we don't import any core api's as it will be built in extension.

dgp1130 replied:
Can we import types? I get we don't want to accidentally pull in our own copy of `@angular/core`, but I think we should be able to:

```typescript
import type { NgGlobal } from '@angular/core';
import type { GetLoadedRoutes } from '@angular/router';
```

Type elision should drop the imports in the JS output since they won't be needed at runtime?

sumitarora committed:
refactor(devtools): making type changes to relaxed

- relaxing typings for GlobalPublish Utils to publish getLoadedRoutes
- fixing pull request comments

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -211,10 +213,15 @@ const getNestedPropertiesCallback =
+  );
+  const rootInjector = (forest[0].resolutionPath ?? []).find((i) => i.name === 'Root');
+  if (rootInjector) {
+    getRouterConfigFromRoot(messageBus)(rootInjector);
```
Suggestion: Is there a need for this currying? Could we just call it as `getRouterConfigFromRoot(messageBus, rootInjector)`?

sumitarora replied:
done

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -340,67 +347,101 @@ function getNodeDIResolutionPath(node: ComponentTreeNode): SerializedInjector[]
+const getRouterConfigFromRoot =
+  (messageBus: MessageBus<Events>) => (injector: SerializedInjector) => {
+    const serializedProviderRecords = getInjectorProvidersCallbackUtil(injector) || [];
+    const routerInstance = serializedProviderRecords.filter(
```
Suggestion: I think `Array.prototype.find` would be more appropriate here? Otherwise we need to assert that we got exactly one router instance.

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -340,67 +347,101 @@ function getNodeDIResolutionPath(node: ComponentTreeNode): SerializedInjector[]
+    const provider = serializedProvider.index.map((index) => providerRecords[index]);
+    return injector.get(provider[0].token, null, {optional: true});
+  } else {
+    return;
```
Question: Should we throw or error in this case? Is this a valid scenario where we have a provider but the index isn't a number or array?

sumitarora replied:
we are using this to get the router data so it would only be called if router provider is there.

dgp1130 commented on devtools/projects/ng-devtools-backend/src/lib/client-event-subscribers.ts:
```diff
@@ -340,67 +347,101 @@ function getNodeDIResolutionPath(node: ComponentTreeNode): SerializedInjector[]
+const getProviderValue = (
+  serializedInjector: SerializedInjector,
+  serializedProvider: SerializedProviderRecord,
+) => {
```
Consider: Related to the previous comment, do we care about disambiguating a provider which returns `undefined` vs not having a provider? The two `return;` statements effectively return `undefined`, but if a provider exists and returns `undefined`, I would get the same result and won't know which case I'm in. Is that a potential issue?

I'm wondering if this should return a discriminated union like:

```typescript
interface NoProvider {
  type: 'no-provider';
}
interface ProviderValue<T> {
  type: 'provider';
  value: T,
}
type ProviderResult<T> = NoProvider | ProviderValue<T>;

function getProviderValue(/* ... */): ProviderResult<unknown> { /* ... */ }
```

WDYT?

dgp1130 commented on packages/core/src/render3/util/global_utils.ts:
```diff
@@ -47,6 +47,10 @@ import {
 
+interface NgGlobalPublishUtils {
+  getLoadedRoutes(route: any): any;
+}
```
Suggestion: I'm thinking we actually shouldn't put a `getLoadedRoutes` type on the global object at all if we have to use `any`. This is public API and so if/when we put real router types on it that would be a breaking change and harder to land.

Instead, I'm thinking we make this private and just leave it off the `ng` global type entirely. `@angular/router` can export the function type directly (`type GetLoadedRoutes = (route: Route) => Route[];`) and DevTools can import and cast to that type directly. We might also want to prefix the property with `ɵ` just to signal that it's internal.

```typescript
import type { ɵGetLoadedRoutes } from '@angular/router';

const ng = globalThis.ng as typeof globalThis.ng & { ɵgetLoadedRoutes?: ɵGetLoadedRoutes };
if (ng.ɵgetLoadedRoutes) {
  const routes = ng.ɵgetLoadedRoutes(route); // Use it.
  // ...
}
```

If we export the type that way, do we need to change `global_utils.ts` at all? Is there any value in the publish internal vs. publish external if we don't go that direction?

sumitarora replied:
It won't work as we can't import anything from router within core. It has to be other way around export from core and import in router.

dgp1130 replied:
This snippet would be in DevTools. DevTools can depend on `@angular/router` and hard cast the `globalThis.ng` type to access a typed `ɵgetLoadedRoutes` function.

dgp1130 reviewed (commented):
Given that some changes were made in review, I'd suggest squashing some of these commits or at least marking them `fixup!` so they get squashed in the merge process.

We should also mark these commits `refactor(devtools)` instead of `feat` since we don't typically include DevTools in the changelog.

@atscott, can you take a look at this from the FW / routing side just to make sure the changes to `@angular/core` and `@angular/router` are reasonable? You probably don't need to worry about the DevTools stuff.

atscott commented on devtools/projects/ng-devtools-backend/src/lib/router-tree.ts:
```diff
@@ -14,55 +14,88 @@ type Routes = any;
 type Router = any;
 
 export function parseRoutes(router: Router): Route {
+  const currentUrl = router?.stateManager?.routerState?.snapshot?.url;
```
Any reason you can't use `router.routerState` here instead of going to the private stateManager property?
